src/mon/OSDMonitor.cc

```diff
@@ -5113,6 +5113,87 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
       rs << "\n";
       rdata.append(rs.str());
     }
+  } else if (prefix == "osd pool application get") {
+    boost::scoped_ptr<Formatter> f(Formatter::create(format, "json-pretty",
+                                                     "json-pretty"));
+    string pool_name;
+    cmd_getval(g_ceph_context, cmdmap, "pool", pool_name);
+    string app;
+    cmd_getval(g_ceph_context, cmdmap, "app", app);
+    string key;
+    cmd_getval(g_ceph_context, cmdmap, "key", key);
+
+    if (pool_name.empty()) {
+      // all
+      f->open_object_section("pools");
+      for (const auto &pool : osdmap.pools) {
+        std::string name("<unknown>");
+        const auto &pni = osdmap.pool_name.find(pool.first);
+        if (pni != osdmap.pool_name.end())
+          name = pni->second;
+        f->open_object_section(name.c_str());
+        for (auto &app_pair : pool.second.application_metadata) {
+          f->open_object_section(app_pair.first.c_str());
+          for (auto &kv_pair : app_pair.second) {
+            f->dump_string(kv_pair.first.c_str(), kv_pair.second);
+          }
+          f->close_section();
+        }
+        f->close_section(); // name
+      }
+      f->close_section(); // pools
+      f->flush(rdata);
+    } else {
+      int64_t pool = osdmap.lookup_pg_pool_name(pool_name.c_str());
+      if (pool < 0) {
+        ss << "unrecognized pool '" << pool_name << "'";
+        r = -ENOENT;
+        goto reply;
+      }
+      auto p = osdmap.get_pg_pool(pool);
+      // filter by pool
+      if (app.empty()) {
+        f->open_object_section("application_metadata");
```

Comment:
There are probably multiple apps associating with one single pool, in which case we still need to be able to distinguish kvs coming from different apps. So I guess we still need this to output the app names here if no app name is specified by user...

For example:

```
ceph osd pool application get test_pool
{
    "fake": {
        "key1": "val1",
        "key2": "val2"
    },
    "rbd": {
        "key1": "val1",
        "key2": "val2",
        "key3": "val3"
    }
}
```


The overall logic here is that we don't repeatedly output the items which have already been specified by user, e.g.:
(1) By specifying nothing, we output all (pool names + app names + kvs):
```
ceph osd pool application get
{
    "cephfs_data_b": {
        "cephfs": {}
    },
    "cephfs_metadata_a": {
        "cephfs": {}
    },
    "test_pool": {
        "rbd": {
            "test": "me"
        }
    }
}
```
(2) By specifying pool name, we output app names + kvs only:
```
ceph osd pool application get test_pool
{
    "rbd": {
        "test": "me"
    }
}
```
(3) By specifying pool name + app name, we output kvs only:

```
ceph osd pool application get test_pool rbd
{
    "test": "me"
}
```
(4) By specifying pool name + app name + key, we output value only:

```
ceph osd pool application get test_pool rbd test
me
```